### PR TITLE
Add new tutorial about docker based CF apps

### PR DESCRIPTION
**Pull request by beyhan**

### CODE REVIEW

File: tutorials/cp-cf-understand-docker-application-lifecycle/cp-cf-understand-docker-application-lifecycle.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+RUN npm install
+#copy over the rest of the app files
+COPY . .
+#the start command to be run
+CMD [ "node", "server.js" ]
+```
+
+In the above code we begin with the official Node image from Docker. We then proceed to add onto this base image, adding files, downloading dependencies, and providing the start command for the `cf-nodejs` app. Make sure to save the file once you have added the code above. Once that is done you then need to tell Docker to build your image. Run the following command (the `username` is your `Dockerhub` username and some commands may require you to enter your `Dockerhub` password):
+
+```
```
Comment (MichaelCzcz) [inline tutorials/cp-cf-understand-docker-application-lifecycle/cp-cf-understand-docker-application-lifecycle.md:63]: Please add label Shell / Bash

File: tutorials/cp-cf-understand-docker-application-lifecycle/cp-cf-understand-docker-application-lifecycle.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+docker build -t <username>/cf-nodejs-app .
+```
+
+You should see output similar to the following:
+
+!![Running docker build](docker-build.png)
+
+Now that the image is built, you need to push it to a repository on `Dockerhub`. Running the following commands, a repository will be created for your user called `cf-nodejs-app` and your image will be uploaded to it:
+
+```
```
Comment (MichaelCzcz) [inline tutorials/cp-cf-understand-docker-application-lifecycle/cp-cf-understand-docker-application-lifecycle.md:74]: Please add label Shell / Bash

File: tutorials/cp-cf-understand-docker-application-lifecycle/cp-cf-understand-docker-application-lifecycle.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+docker login -u <docker-id>
+docker push <username>/cf-nodejs-app
+```
+
+!![Pushing the Docker image](docker-push.png)
+
+Finally, once the image has been uploaded, you can push the image to Cloud Foundry (you can replace the name `docker-nodejs` with whatever you want to name your application).
+
+```
```
Comment (MichaelCzcz) [inline tutorials/cp-cf-understand-docker-application-lifecycle/cp-cf-understand-docker-application-lifecycle.md:83]: Please add label Shell / Bash

File: tutorials/cp-cf-understand-docker-application-lifecycle/cp-cf-understand-docker-application-lifecycle.md [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+cf push docker-nodejs --docker-image <username>/cf-nodejs-app:latest --docker-username <username>
+```
+
+> In this example `cf-nodejs-app:latest` is used, but in the next section you will see why using the 'latest' image may not always be what you want.
+
+Finally, once the application has started you should be able to access the app through a browser and see the same output as when you ran the `cf-nodejs` app using a buildpack. You can also validate that your application is actually using the docker image by running:
+
+```
```
Comment (MichaelCzcz) [inline tutorials/cp-cf-understand-docker-application-lifecycle/cp-cf-understand-docker-application-lifecycle.md:91]: Please add label Shell / Bash

File: tutorials/cp-cf-understand-docker-application-lifecycle/cp-cf-understand-docker-application-lifecycle.md [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+COPY package*.json ./
+#install app dependencies
+RUN npm install
+#copy over the rest of the app files
+COPY . .
+#the start command to be run
+CMD [ "node", "server.js" ]
+```
+
+In the above code we begin with the official Node image from Docker. We then proceed to add onto this base image, adding files, downloading dependencies, and providing the start command for the `cf-nodejs` app. Make sure to save the file once you have added the code above. Once that is done you then need to tell Docker to build your image. Run the following command (the `username` is your `Dockerhub` username and some commands may require you to enter your `Dockerhub` password):
```
Comment (MichaelCzcz) [inline tutorials/cp-cf-understand-docker-application-lifecycle/cp-cf-understand-docker-application-lifecycle.md:61]: Please don't use the "royal we" when you are giving instructions to the user.

File: tutorials/cp-cf-understand-docker-application-lifecycle/cp-cf-understand-docker-application-lifecycle.md [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+  4. Related to the above, buildpacks are generally maintained and updated by the Cloud Foundry community, but this is not the case for referenced Docker base images. Here you will need to keep track of security vulnerabilities and update the base images. If a base image is not maintained any more you will need to maintain your own base images. Moral of the story: don't take on more than you absolutely have to, but be certain you trust the source of all your base images and buildpacks.
+
+[VALIDATE_1]
+
+[ACCORDION-END]
+
+[ACCORDION-BEGIN [Step 2 ](Push a Docker based application)]
+
+To see how docker based applications work you will take the [cf-nodejs](https://github.com/SAP-archive/cf-sample-app-nodejs) application you deployed before and create a docker image of it, which you will then deploy. To do this, the first thing you need to do is create a `Dockerfile` in the same directory as your application files. The path to the `Dockerfile` should be `<path-to-cf-sample-app-nodejs>/Dockerfile` Once the file is created, open the `Dockerfile` with your favorite editor, and add the following (lines beginning with `#` are comments explaining what the line below it does):
```
Comment (MichaelCzcz) [inline tutorials/cp-cf-understand-docker-application-lifecycle/cp-cf-understand-docker-application-lifecycle.md:44]: Docker is spelled with an upper case D

File: tutorials/cp-cf-understand-docker-application-lifecycle/cp-cf-understand-docker-application-lifecycle.md [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+  1. When you `cf push` with a Docker image, the last step of the process is to schedule the app as a long running process, and at this point the Docker image is downloaded. That means that Cloud Foundry will require access to the container registry where the image is stored to start your application.
+
+  2. Access to the registry is also required when the application needs to be re-staged or restarted. If the registry is not accessible, these operations will fail. For this reason it is required to re-push your application if the credentials for a private container registry change.
+
+  3. Additionally, when you use a buildpack, your main concern is your application code. The buildpack takes care of the work of downloading dependencies, compiling code, and creating a droplet from which your application can run. However, when using a docker image, you have to do all this yourself (set up the image, define dependencies, provide settings the image/app needs to start, etc.)
```
Comment (MichaelCzcz) [inline tutorials/cp-cf-understand-docker-application-lifecycle/cp-cf-understand-docker-application-lifecycle.md:34]: Docker is spelled with an upper case D

File: tutorials/cp-cf-understand-docker-application-lifecycle/cp-cf-understand-docker-application-lifecycle.md [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+Finally, once the image has been uploaded, you can push the image to Cloud Foundry (you can replace the name `docker-nodejs` with whatever you want to name your application).
+
+```
+cf push docker-nodejs --docker-image <username>/cf-nodejs-app:latest --docker-username <username>
+```
+
+> In this example `cf-nodejs-app:latest` is used, but in the next section you will see why using the 'latest' image may not always be what you want.
+
+Finally, once the application has started you should be able to access the app through a browser and see the same output as when you ran the `cf-nodejs` app using a buildpack. You can also validate that your application is actually using the docker image by running:
```
Comment (MichaelCzcz) [inline tutorials/cp-cf-understand-docker-application-lifecycle/cp-cf-understand-docker-application-lifecycle.md:89]: Docker is spelled with an upper case D

File: tutorials/cp-cf-understand-docker-application-lifecycle/cp-cf-understand-docker-application-lifecycle.md [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+
+```
+cf app docker-nodejs
+```
+
+You should then see the `docker image` field with the name your docker image next to it, similar to the output below.
```
Comment (MichaelCzcz) [inline tutorials/cp-cf-understand-docker-application-lifecycle/cp-cf-understand-docker-application-lifecycle.md:95]: Docker is spelled with an upper case D

File: tutorials/cp-cf-understand-docker-application-lifecycle/cp-cf-understand-docker-application-lifecycle.md (hunk 1 above)
Comment (beyhan) [reply tutorials/cp-cf-understand-docker-application-lifecycle/cp-cf-understand-docker-application-lifecycle.md]: Done

File: tutorials/cp-cf-understand-docker-application-lifecycle/cp-cf-understand-docker-application-lifecycle.md (hunk 2 above)
Comment (beyhan) [reply tutorials/cp-cf-understand-docker-application-lifecycle/cp-cf-understand-docker-application-lifecycle.md]: Done

File: tutorials/cp-cf-understand-docker-application-lifecycle/cp-cf-understand-docker-application-lifecycle.md (hunk 3 above)
Comment (beyhan) [reply tutorials/cp-cf-understand-docker-application-lifecycle/cp-cf-understand-docker-application-lifecycle.md]: Done

File: tutorials/cp-cf-understand-docker-application-lifecycle/cp-cf-understand-docker-application-lifecycle.md (hunk 4 above)
Comment (beyhan) [reply tutorials/cp-cf-understand-docker-application-lifecycle/cp-cf-understand-docker-application-lifecycle.md]: Done

File: tutorials/cp-cf-understand-docker-application-lifecycle/cp-cf-understand-docker-application-lifecycle.md (hunk 5 above)
Comment (beyhan) [reply tutorials/cp-cf-understand-docker-application-lifecycle/cp-cf-understand-docker-application-lifecycle.md]: Done

File: tutorials/cp-cf-understand-docker-application-lifecycle/cp-cf-understand-docker-application-lifecycle.md (hunk 6 above)
Comment (beyhan) [reply tutorials/cp-cf-understand-docker-application-lifecycle/cp-cf-understand-docker-application-lifecycle.md]: Done

File: tutorials/cp-cf-understand-docker-application-lifecycle/cp-cf-understand-docker-application-lifecycle.md (hunk 7 above)
Comment (beyhan) [reply tutorials/cp-cf-understand-docker-application-lifecycle/cp-cf-understand-docker-application-lifecycle.md]: Done

File: tutorials/cp-cf-understand-docker-application-lifecycle/cp-cf-understand-docker-application-lifecycle.md (hunk 8 above)
Comment (beyhan) [reply tutorials/cp-cf-understand-docker-application-lifecycle/cp-cf-understand-docker-application-lifecycle.md:89]: Done

File: tutorials/cp-cf-understand-docker-application-lifecycle/cp-cf-understand-docker-application-lifecycle.md (hunk 9 above)
Comment (beyhan) [reply tutorials/cp-cf-understand-docker-application-lifecycle/cp-cf-understand-docker-application-lifecycle.md]: Done

**Pull request merged**